### PR TITLE
feat: introduces local artifact server and materialize subcommand

### PR DESCRIPTION
**Pull request by ricardomaraschini**
- implement and install a localhost http server for local assets.
- implement a hidden 'materialise' command to materialise the embedded assets into a directory.

### CODE REVIEW

File: cmd/embedded-cluster/install.go [hunk 1]
```diff
@@ -52,6 +51,30 @@ func runCommand(bin string, args ...string) (string, error) {
 	return stdout.String(), nil
 }
 
+// installAndEnableLocalArtifactMirror installs and enables the local artifact mirror. This
+// service is reponsible for serving on localhost, through http, all files that are used
+// during a cluster upgrade.
+func installAndEnableLocalArtifactMirror() error {
```
Comment (laverya) [inline cmd/embedded-cluster/install.go:57]: this needs to be called from join too

File: cmd/embedded-cluster/materialise.go [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+	"github.com/replicatedhq/embedded-cluster/pkg/goods"
+)
+
+var materialiseCommand = &cli.Command{
+	Name:      "materialise",
+	Usage:     "Materialise embedded assets into a directory",
+	Hidden:    true,
+	UsageText: fmt.Sprintf("%s materialise <dir>", defaults.BinaryName()),
+	Before: func(c *cli.Context) error {
+		if len(c.Args().Slice()) != 1 {
```
Comment (laverya) [inline cmd/embedded-cluster/materialise.go:18]: check for root?

File: cmd/local-artifact-mirror/main.go [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+		handler.ServeHTTP(w, r)
+	})
+}
+
+// serveCommand starts a http server that serves files from the /var/lib/embedded-cluster
+// directory. This servers listen only on localhost and is used to serve files needed by
+// the autopilot during an upgrade.
+var serveCommand = &cli.Command{
+	Name:  "serve",
+	Usage: "Serve local files over HTTP",
```
Comment (laverya) [inline cmd/local-artifact-mirror/main.go:56]: this should include the path IMO

File: e2e/local-artifact-mirror_test.go [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+
+	commands := [][]string{
+		{"apt-get", "install", "curl", "-y"},
+		{"systemctl", "status", "local-artifact-mirror"},
+		{"systemctl", "stop", "local-artifact-mirror"},
+		{"systemctl", "start", "local-artifact-mirror"},
+		{"systemctl", "status", "local-artifact-mirror"},
+		{"curl", "-o", "/tmp/kubectl-test", "127.0.0.1:50000/bin/kubectl"},
+		{"chmod", "755", "/tmp/kubectl-test"},
+		{"/tmp/kubectl-test", "version", "--client"},
```
Comment (laverya) [inline e2e/local-artifact-mirror_test.go:37]: some things I would like to see tested here:
1. does it properly block access to logs?
2. what happens if you include a few `../../`s in the path?

File: e2e/utils.go [hunk 5]
```diff
@@ -24,9 +24,13 @@ func RunCommandsOnNode(t *testing.T, cl *cluster.Output, node int, cmds [][]stri
...
 		t.Logf("running `%s` node %d", cmdstr, node)
 		stdout, stderr, err := RunCommandOnNode(t, cl, node, cmd)
-		if err != nil {
+		if len(stdout) > 0 || len(stderr) > 0 {
+			t.Logf("-----")
 			t.Logf("stdout:\n%s", stdout)
 			t.Logf("stderr:\n%s", stderr)
+			t.Logf("-----")
+		}
+		if err != nil {
```
Comment (laverya) [inline e2e/utils.go:33]: if you do this you should make sure that everything that calls RunCommandsOnNode isn't logging duplicates

File: pkg/defaults/provider.go [hunk 6]
```diff
@@ -17,20 +17,16 @@ import (
 // Base is the base directory inside which all the other directories are
 // created.
 func NewProvider(base string) *Provider {
-	obj := &Provider{Base: base}
-	obj.Init()
-	return obj
+	return &Provider{Base: base, Home: "/var/lib/embedded-cluster"}
```
Comment (laverya) [inline pkg/defaults/provider.go:20]: I am very confused by what you are doing in this file

File: pkg/goods/goods.go [hunk 7]
```diff
@@ -83,3 +83,19 @@ func Materialize() error {
 	}
 	return nil
 }
+
+//go:embed systemd/*
+var systemdfs embed.FS
+
+// MaterializeLocalArtiactMirrorUnitFile writes to disk the local-artifact-mirror systemd unit file.
+func MaterializeLocalArtiactMirrorUnitFile() error {
```
Comment (laverya) [inline pkg/goods/goods.go:91]:
```suggestion
// MaterializeLocalArtifactMirrorUnitFile writes to disk the local-artifact-mirror systemd unit file.
func MaterializeLocalArtifactMirrorUnitFile() error {
```

File: pkg/defaults/provider.go (hunk 6 above)
Comment (laverya) [reply pkg/defaults/provider.go:20]: I suppose it's OK

File: pkg/defaults/defaults.go [hunk 8]
```diff
@@ -14,8 +14,8 @@ var (
 	TroubleshootVersion = "0.0.0"
 	// KubectlVersion holds the version of kubectl binary we are embedding.
 	KubectlVersion = "0.0.0"
-	// provider holds a global reference to the default provider.
-	provider *Provider
+	// GlobalProvider holds a global reference to the default provider.
+	GlobalProvider *Provider
```
[the remark this answers is not in the document]
Comment (ricardomaraschini) [reply pkg/defaults/defaults.go:18]: We are overwriting it on `materialise` so we can materialise things into a different directory.

File: pkg/defaults/provider.go (hunk 6 above)
Comment (ricardomaraschini) [reply pkg/defaults/provider.go:20]: Yeah, when running this from within a pod we may want to materialise the files into a different location. I don't like this anymore, let me revisit.

File: e2e/utils.go (hunk 5 above)
Comment (ricardomaraschini) [reply e2e/utils.go:33]: we are not logging duplicates.

File: e2e/local-artifact-mirror_test.go (hunk 4 above)
Comment (ricardomaraschini) [reply e2e/local-artifact-mirror_test.go:37]: done.

File: pkg/defaults/defaults.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (ricardomaraschini) [reply pkg/defaults/defaults.go:18]: This relates to https://github.com/replicatedhq/embedded-cluster/pull/443#discussion_r1521791505 so I am resolving this. We can discuss there.

File: pkg/defaults/provider.go (hunk 6 above)
Comment (ricardomaraschini) [reply pkg/defaults/provider.go]: OK, this has been removed from the PR as now we require `root` and materialise directly to `/var/lib/embedded-cluster`.

**Review by laverya — APPROVED** [review]

**Pull request merged**